Caroll/smart apis blog posts (#96)
vendia/blog: posts/2022-04-04-smart-apis.md

date: '2022-04-04'
authors:
  - Tim Wagner
---

## **APIs: A Tradition of “Dumb Pipes”** 🔧

APIs are critical to virtually all modern applications. From internal APIs shared between teams or departments to the surface area of global enterprises, APIs represent the public “contracts” of modern software systems and IT solutions. 

API protocols have also matured over the years, from [Electronic Data Interchange (EDI)](https://www.edibasics.com/what-is-edi/) to HTTP to [GraphQL](https://graphql.org/learn/), as both businesses and developers have evolved their demands on the data that APIs need to carry. The infrastructure for implementing APIs has advanced as well, from “rolling your own” to deploying open source libraries (such as [NGINX](https://en.wikipedia.org/wiki/Nginx)), to fully managed cloud services such as [Amazon API Gateway](https://aws.amazon.com/api-gateway/) and [Azure API Management](https://azure.microsoft.com/en-us/services/api-management/). 

But for all these advancements, the conceptual role of an API – that of a “dumb pipe” that doesn’t understand, control, or mediate its content – has remained largely unchanged: APIs are all “syntax”, no “semantics”. 

> To put it another way, *a**n API doesn’t care if the data it carries (and carrying data is largely what APIs do) is right or wrong, consistent or inconsistent, complete or partial, up to date or radically out of date.*** And even if the data you drop into one side is perfect, by the time it reaches someone’s system on the other side it may no longer be.

Reviewer: ```suggestion
To put it another way, ***an API doesn’t care if the data it carries (and carrying data is largely what APIs do) is right or wrong, consistent or inconsistent, complete or partial, up to date or radically out of date.*** And even if the data you drop into one side is perfect, by the time it reaches someone’s system on the other side it may no longer be.
```

## The Cost of Dumb Pipes

> As a result, a ton of developer time (and company 💰) go into trying to close what we like to call the API data gap. And that gap is huge: Sharing data, and keeping it consistent across companies, clouds, geographies (aka regions), accounts, and technology stacks is arguably the single hardest challenge in modern day applications. 

Reviewer: ```suggestion
As a result, a ton of developer time (and company 💰) go into trying to close what we like to call the API data gap. And that gap is huge: sharing data, and keeping it consistent across companies, clouds, geographies (aka regions), accounts, and technology stacks is arguably the single hardest challenge in modern day applications. 
```
